src/package.c

```diff
@@ -975,18 +975,70 @@ int apk_ipkg_add_script(struct apk_installed_package *ipkg,
 	return 0;
 }
 
-void apk_ipkg_run_script(struct apk_installed_package *ipkg,
-			 struct apk_database *db,
-			 unsigned int type, char **argv)
+static int run_script(char** argv, int dirfd, char* fn)
 {
+	int status;
+	pid_t pid;
 	static char * const environment[] = {
 		"PATH=/usr/sbin:/usr/bin:/sbin:/bin",
 		NULL
 	};
+
+	pid = fork();
+	if (pid == -1) {
+		apk_error("fork() failed with error str: %s", apk_error_str(errno));
+		return -1;
+	}
+	if (pid == 0) {
+		umask(0022);
+		if (fchdir(dirfd) == 0) {
+			if (fn == NULL)
+				execve(argv[0], argv, environment); /* pre/post hook */
+			else if (chroot(".") == 0)
+				execve(fn, argv, environment); /* apk "internal" script */
+			else
+				apk_error("chroot() failed for script with error str: %s", apk_error_str(errno));
+			apk_error("execve() failed for script with error str: %s", apk_error_str(errno));
+		} else
+			apk_error("change working dir failed with error str: %s", apk_error_str(errno));
+		exit(1); /* should not get here */
+	}
+	waitpid(pid, &status, 0);
+	if (!WIFEXITED(status) || WEXITSTATUS(status) != 0) {
+		apk_error("script exited with error %d", WEXITSTATUS(status));
+		return -1;
+	}
+	return 0;
+}
+
+int apk_run_hook_script(void *ctx, int dirfd, const char *file)
+{
+	char fn[PATH_MAX];
+	char *stage = (char *)ctx;
+	char *script_args[] = { "/bin/sh", "-c", "", NULL };
+
+	if ((apk_flags & (APK_NO_SCRIPTS | APK_SIMULATE)) != 0)
+		return 0;
+
+	snprintf(fn, sizeof(fn), "./" "%s", file);
+	script_args[2] = (char *) fn;
+
+	if (apk_verbosity >= 2)
+		apk_message("Calling apk %s-script: %s\n", stage, fn);
+
+	if (run_script(script_args, dirfd, NULL) != 0)
+		exit(1);
```

Comment:
Is this exit() intentional? It can abort apk during commit, and not clean up things. This is especially nasty thing to do in the post script where files are left unpacked, but database is probably uncommitted.